Enhance Discord Bot Notifications for Transcript Queuing

nully0x opened the pull request:
This PR addresses #250 and #253 since they are minor update

- refactored transcript Url to link to btctranscript
- modified message format
- handle multiple commits with an extra catch for processing error
- create a migration to remove the unique constraint on transcriptHash (will need to run migration locally)

kouloumos commented on api/app/helpers/sendAlert.ts:
```diff
@@ -7,51 +7,45 @@ function transformUrl(transcriptUrl?: string | null): string | null {
-    content = `🤖 New Transcript Ready for Review!\nTitle: ${transcriptTitle}\nSpeakers: ${speakers}\nLink: ${transformedUrl}`;
-  }
+  // Use template literals for constructing content
+  const content = isError ? `Error: ${message}` : `🤖 ${message}\nTitle: ${transcriptTitle}\nSpeakers: ${speakers}\nLink: ${transformedUrl}`;
```
let's add a message title for when we have an error. Maybe:
"❌ Transcript not added to the Queue"

kouloumos commented on api/app/controllers/webhook.controller.ts:
```diff
@@ -319,16 +342,29 @@ export async function handlePushEvent(req: Request, res: Response) {
+        try {
+          await processCommit(commit, pushEvent, "added", branch);
+          await processCommit(commit, pushEvent, "modified", branch);
+          processedCommits++;
```
If we have commit A,B,C and B throws an error inside of `processCommit` then `processedCommits` will be equal to 1, but B was also processed - its processing resulted in an error. So the correct number is 2 processed commits, therefore 1 unprocessed.

Extheoisah commented on api/app/helpers/sendAlert.ts:
```diff
@@ -1,27 +1,52 @@
+  const webhookUrl = isError ? process.env.DISCORD_ERROR_WEBHOOK_URL : process.env.DISCORD_TRANSCRIPT_WEBHOOK_URL;
+
+  if (!webhookUrl) {
+    console.error("Webhook URL is not set");
```
Let's use the logger util function instead of console.error

kouloumos commented on api/app/helpers/sendAlert.ts:
```diff
@@ -1,27 +1,54 @@
   }
 
-  const content = `New Queuer Action\nMessage: ${message}\nTranscript Title: ${transcriptTitle}\nTranscriptUrl: ${transcriptUrl}\ntranscriptHash: ${transcriptHash}`;
+  const transformedUrl = transformUrl(transcriptUrl);
```
When `sendAlert` is called with an error (`isError = true`) there is no `transcriptUrl`, so `transformUrl` always fails.

You can just skip this calculation when there is an error.

Extheoisah commented on api/app/controllers/webhook.controller.ts:
```diff
@@ -174,109 +178,120 @@ async function processCommit(
-    const transcriptHash = generateUniqueHash(jsonContent);
-    const totalWords = getTotalWords(jsonContent.body);
-    const content = jsonContent;
+      if (!jsonContent) {
```
We should do this check on line 188 before creating the `transcript_by` variable

Extheoisah commented on api/app/controllers/webhook.controller.ts:
```diff
@@ -174,109 +178,120 @@ async function processCommit(
+        parseMdToJSON<BaseParsedMdContent>(mdContent);
+      const transcript_by = jsonContent.transcript_by.toLowerCase();
+
+      function isTranscriptValid(transcript_by: string): boolean {
```
this function should be in utils for transcripts not defined inline the loop.

nully0x replied:
fixed

Extheoisah replied:
this hasn't been fixed. can't find the commit that fixes it

Extheoisah commented on api/app/controllers/webhook.controller.ts:
```diff
@@ -174,109 +178,120 @@ async function processCommit(
-      status: TranscriptStatus.queued,
-      contentTotalWords: totalWords,
-    };
+      if (!isTranscriptValid(transcript_by)) {
```
We should perform this check after we confirm that we have valid JSON content and before we make any db call.

nully0x replied:
fixed

Extheoisah commented on api/app/controllers/webhook.controller.ts:
```diff
@@ -174,109 +178,120 @@ async function processCommit(
+        contentTotalWords: totalWords,
+      };
+
+      let transcriptData: any;
```
why type `any` and not `TranscriptAttributes`?

nully0x replied:
Quite an intricacy for me here, cos first in the `TranscriptAttributes` interface, transcriptUrl is `string | null`,  wanted to change it to `string | undefined `to be able to use it in the sending alert logic `line 268`. another option is to make transcriptUrl option(don't want to do that, cos its critical for further operation)

Extheoisah replied:
Not sure I understand this but you can type cast the transcriptUrl to what you want when passing it to the `sendAlert` fn. Ideally, we should have one type for the url

Extheoisah commented on api/app/controllers/webhook.controller.ts:
```diff
@@ -174,109 +178,120 @@ async function processCommit(
+
+      let transcriptData: any;
+
+      async function cacheTranscript(transcriptData: any) {
```
Let's make this a reusable function in utils for transcript. We're doing the same thing here https://github.com/bitcointranscripts/transcription-review-backend/blob/b5e7e32801a806d8ae2234f9b21ffccf6958b1e5/api/app/controllers/transcript.controller.ts#L58 from line 58 to line 78

nully0x replied:
Yea, we raised in chats last week, I still left it here because it's a WIP, will move it out

Extheoisah commented on api/app/controllers/webhook.controller.ts:
```diff
@@ -318,17 +341,19 @@ export async function handlePushEvent(req: Request, res: Response) {
+    return handleError(error, res);
   }
-  return res.status(200).json({ message: "Transcript queued Successfully" });
+  return res.sendStatus(200);
```
any reason why we changed this?

nully0x replied:
This is what is sent back to the GitHub webhook portal. I don't see it, if it runs well, 200 OK is enough response(IMO), if there is an error the error and its message are sent.

Extheoisah replied:
ok

Extheoisah commented on api/app/helpers/sendAlert.ts:
```diff
@@ -1,27 +1,59 @@
   transcriptTitle?: string | null,
-  transcriptUrl?: string | null,
-  transcriptHash?: string | null
+  speakers?: string | null,
```
since this is optional no need to explicitly define the type of null
```suggestion
  speakers?: string,
```

Extheoisah commented on api/app/helpers/sendAlert.ts:
```diff
@@ -1,27 +1,59 @@
-  transcriptUrl?: string | null,
-  transcriptHash?: string | null
+  speakers?: string | null,
+  transcriptUrl?: string | null
```
```suggestion
  transcriptUrl?: string
```

nully0x commented:
I addressed possible bottleneck we might run into while queueing and sending alerts for large data in this  [ commit ](https://github.com/bitcointranscripts/transcription-review-backend/pull/256/commits/189d125ea34d682b51cca934755771d1d1814320)

Initaillly, while queing transcripts there is no issue with commits < 5 but as we are getting to merge I did performance testing and noticed processing  commits > 7 takes ~10sec which means we won't be able to queue up to 10 transcript at a go (pushEvent) because github webhook expects to get a response while processing in 10sec, to which the implementation timed out for 10 commits while testing bechmarked at  15sec (varies on content of the commit). 

before 
![image](https://github.com/bitcointranscripts/transcription-review-backend/assets/40327060/70c54706-2e87-4646-85d6-9f9da5dfd472)

now 
![image](https://github.com/bitcointranscripts/transcription-review-backend/assets/40327060/6c28e9ff-635e-425f-854d-7f04372ef6ad)

by this we should be able to process ~ 40 commits in one push request.

Second issue was sending alert for the commit actions in the push event. Discord raises a 429(surpassing rate limit when sending up to 10 alert at the same time) but does really stop sending but leads to a 403(unauthorized) error. This won't allow us know what is wrong with subsequent commit  in the same event. 

To mitigate this, I implemented a bottleneck to delay sending each request per 3 seconds to avoid running into rate limiting with discord

To test: please run yarn install to refresh you packages.

kouloumos reviewed (approved):
Tested with multiple commits and looks good!
